pinot/regressors/biophysical_regressor.py

```diff
@@ -0,0 +1,78 @@
+# =============================================================================
+# IMPORTS
+# =============================================================================
+import torch
+import pinot
+import abc
+import math
+import numpy as np
+
+class BiophysicalRegressor(torch.nn.Module):
+    r""" Biophysically inspired model
+
+    Parameters
+    ----------
+
+    log_sigma : `float`
+        ..math:: \log\sigma observation noise
+
+    base_regressor : a regressor object that generates a latent F
+
+
+    """
+
+    def __init__(self, base_regressor=None):
+        super(BiophysicalRegressor, self).__init__()
+        self.base_regressor = base_regressor
+        self.log_sigma_measurement = torch.nn.Parameter(torch.zeros(1))
+
+
+	def g(self, func_value=None, test_ligand_concentration=1e-3):
+	    return 1 / (1 + torch.exp(-func_value) / test_ligand_concentration)
+
+
+	def condition(self, h=None, test_ligand_concentration=1e-3, **kwargs):
+		distribution_base_regressor = self.base_regressor.condition(h, **kwargs)
+		#we sample from the latent f to push things through the likelihood
+		#Note: if we do this, in order to get good estimates of LLK we may need to draw multiple samples
+		f_sample = distribution_base_regressor.rsample()
+		mu_m = self.g(func_value=f_sample, test_ligand_concentration=test_ligand_concentration)
+		sigma_m = torch.exp(self.log_sigma_measurement)
+		distribution_measurement = torch.distributions.normal.Normal(loc=mu_m, scale=sigma_m)
+		return distribution_measurement
+
+
+	def loss(self, h=None, y=None, test_ligand_concentration=1e-3, **kwargs):
+		distribution_measurement = self.condition(h=h, test_ligand_concentration=test_ligand_concentration, **kwargs)
+		loss_measurement = -distribution_measurement.log_prob(y)
+		return loss_measurement
+
+
+   	def marginal_sample(self, h=None, n_samples=100, test_ligand_concentration=1e-3, **kwargs):
+   		distribution_base_regressor = self.base_regressor.condition(h, **kwargs)
+   		samples_measurement = []
+   		for ns in range(n_samples):
+   			f_sample = distribution_base_regressor.rsample()
+	        mu_m = self.g(func_value=f_sample, test_ligand_concentration=test_ligand_concentration)
+	        sigma_m = torch.exp(self.log_sigma_measurement)
+	        distribution_measurement = torch.distributions.normal.Normal(loc=mu_m, scale=sigma_m)
+    		samples_measurement.append(distribution_measurement.sample())
```

Comment:
ok the source of my confusion is that these lines need to be in the for-loop

Reply:
technically ```sigma_m = torch.exp(self.log_sigma_measurement)``` does not need to be in there, but the others do. I just left it there to group the parameters for the measurement distribution.

The reason is: the mean is being re-calculated every time we sample a new deltaG

Reply:
confusion resolved?

Reply:
no, these need to be moved a tab forward, inside the for-loop. that's what I'm referring to 😀

otherwise you'll end up with len(samples_measurement) == 1 and it contains the distribution_measurement.sample() corresponding to mu_m from just one f_sample

Reply:
but fair enough, you could do something like:

```
   	        sigma_m = torch.exp(self.log_sigma_measurement)
		samples_measurement = []
   		for ns in range(n_samples):
   			f_sample = distribution_base_regressor.rsample()
	        	mu_m = self.g(func_value=f_sample, test_ligand_concentration=test_ligand_concentration)
	        	distribution_measurement = torch.distributions.normal.Normal(loc=mu_m, scale=sigma_m)
    			samples_measurement.append(distribution_measurement.sample())
```

Reply:
I agree it's not as nice-looking though

Reply:
wait, in my code they ARE inside the for loop...

Reply:
<img width="751" alt="Screen Shot 2020-07-01 at 10 17 54 PM" src="https://user-images.githubusercontent.com/1401135/86319199-d1823680-bbe8-11ea-9b30-b305fd0d4e5d.png">

Reply:
what I see:

![image](https://user-images.githubusercontent.com/497074/86319206-fafb8c00-bc01-11ea-997a-ad83c281b7d8.png)

Reply:
probably another weird tab thing..

Reply:
ok, that explains our differences. Let me check this.

Reply:
in GitHub I also see f_sample as an extra tab forward instead of left-aligned with everything else

Reply:
check it all now

Reply:
oh man, my text editor shows the code formatted very differently than here. 
I'll check its indent options